crypto/asn1/d2i_pr.c

```diff
@@ -15,15 +15,70 @@
 #include <openssl/bn.h>
 #include <openssl/evp.h>
 #include <openssl/objects.h>
+#include <openssl/decoder.h>
 #include <openssl/engine.h>
 #include <openssl/x509.h>
 #include <openssl/asn1.h>
 #include "crypto/asn1.h"
 #include "crypto/evp.h"
+#include "internal/asn1.h"
 
-EVP_PKEY *d2i_PrivateKey_ex(int type, EVP_PKEY **a, const unsigned char **pp,
+static const char *pkey_type_to_encoder_name(int type)
+{
+    switch (type) {
+    case EVP_PKEY_RSA:
+        return "RSA";
+    case EVP_PKEY_RSA_PSS:
+        return "RSA-PSS";
+    case EVP_PKEY_DH:
+        return "DH";
+    case EVP_PKEY_DHX:
+        return "DHX";
+    case EVP_PKEY_EC:
+        return "EC";
+    case EVP_PKEY_DSA:
+    case EVP_PKEY_X25519:
+    case EVP_PKEY_ED25519:
+    case EVP_PKEY_X448:
+    case EVP_PKEY_ED448:
+        return OBJ_nid2sn(type);
+    default:
+        return NULL; /* Not supported */
+    }
+}
+
+EVP_PKEY *d2i_PrivateKey_ex(int keytype, EVP_PKEY **a, const unsigned char **pp,
                             long length, OSSL_LIB_CTX *libctx,
                             const char *propq)
+{
+    OSSL_DECODER_CTX *dctx = NULL;
+    size_t len = length;
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY **ppkey = &pkey;
+    const char *input_type = NULL;
+
+    if (keytype != EVP_PKEY_NONE) {
+        input_type = pkey_type_to_encoder_name(keytype);
+        if (input_type == NULL)
+            return NULL;
+    }
+    if (a != NULL && *a != NULL)
+        ppkey = a;
+
+    dctx = OSSL_DECODER_CTX_new_by_EVP_PKEY(ppkey, "DER", NULL, NULL,
+                                            EVP_PKEY_KEYPAIR, libctx, propq);
+    if (!OSSL_DECODER_from_data(dctx, pp, &len)) {
+        if (ppkey != a)
+            EVP_PKEY_free(*ppkey);
+        pkey = NULL;
+    }
+
```

Comment:
Do I need to pull the private key out here to check that it worked? The selection is basically 'all' so maybe this could just return a public key?

Reply:
You could use `evp_keymgmt_has` with the selection `OSSL_KEYMGMT_SELECT_PRIVATE_KEY` to check...  and yes, I believe that would be a good idea.  But in that case, don't pass `pp` directly to `OSSL_DECODER_from_data()`